cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringStore.java

```diff
@@ -79,6 +79,28 @@ public void addPeer(PeerInfo peerInfo) throws IOException, PeerAlreadyExistsExce
     }, IOException.class, PeerAlreadyExistsException.class);
   }
 
+  /**
+   * Updates peer information.
+   *
+   * @param peerInfo peer information
+   * @throws IOException if inserting into the table fails
+   */
+  public void updatePeer(PeerInfo peerInfo) throws IOException {
```

Comment:
it seems simpler to modify (and rename) addPeer() so that it doesn't throw a PeerAlreadyExistsException and just always upserts without bothering to do a read first. Or is there some other place where we want to be able to tell if the peer already exists?

Reply:
- We want addPeer() to fail on the client side if a tethering connection to the server already exists - https://github.com/cdapio/cdap/blob/fb6ab30742db324703cf11ccd946301f77cb5966/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringClientHandler.java#L76
- I think we'll still need to do a read on the server side. If the peer doesn't already exist, we want to add it in PENDING state. If it does exist, then we want to update all the PeerInfo fields except for the tethering status. e.g: if a tethering connection was in ACCEPTED state, and then client recreates tethering with a different list of namespaces, we want the connection to be automatically accepted on the server side with the updated list of namespaces

Reply:
Fixed